src/py/flwr/simulation/ray_transport/ray_actor.py

```diff
@@ -0,0 +1,280 @@
+            )
+            raise ClientException(message) from ex
+
+        return client_id, client_results
+
+
+def pool_size_from_resources(client_resources: Dict):
+    """Calculate number of Actors that fit in pool given the resources in the.
+
+    cluster and those required per client.
+    """
+    cluster_resources = ray.cluster_resources()
+    num_cpus = cluster_resources["CPU"]
+    num_gpus = cluster_resources.get("GPU", 0)  # there might not be GPU
+    num_actors = int(num_cpus / client_resources["num_cpus"])
+    # if a GPU is present and client resources do require one
+    if client_resources["num_gpus"] > 0.0:
+        if num_gpus:
+            # if there are gpus in the cluster
+            num_actors = min(num_actors, int(num_gpus / client_resources["num_gpus"]))
+        else:
+            num_actors = 0
+
+    if num_actors == 0:
+        log(
+            WARNING,
+            f"Your ActorPool is empty. Your system ({num_cpus = }, {num_gpus = }) "
+            "does not meet the criteria to host at least one client with resources:"
+            f" {client_resources}. Consider lowering your `client_resources`",
+        )
+
+    return num_actors
+
+
+class VirtualClientEngineActorPool(ActorPool):
+    """A pool of VirtualClientEngine Actors."""
+
+    def __init__(self, client_resources: Dict, max_restarts: int = 1):
+        self.client_resources = client_resources
+        self.actor_max_restarts = max_restarts
+        num_actors = pool_size_from_resources(client_resources)
+        actors = [
+            VirtualClientEngineActor.options(
+                **client_resources, max_restarts=max_restarts
+            ).remote()
+            for _ in range(num_actors)
+        ]
+
+        super().__init__(actors)
+
+        self._cid_to_future = {}  # a dict
+        self.actor_to_remove: Set[str] = set()  # a set
+        self.num_actors = len(actors)
+
+        self.lock = threading.RLock()
+
+        # TODO: asyncio check every N seconds if cluster has grown
+        # --> add more actors to the pool if so
+
+    def __reduce__(self):
+        """Make this class serialisable (needed due to lock)."""
+        return VirtualClientEngineActorPool, (
+            self.client_resources,
+            self.actor_max_restarts,
+        )
+
+    def submit(self, fn: Any, value: Callable, cid: str) -> None:
+        """Take idle actor and assign it a client workload."""
+        actor = self._idle_actors.pop()
+        if self._check_and_remove_actor_from_pool(actor):
+            future = fn(actor, value)
+            future_key = tuple(future) if isinstance(future, List) else future
+            self._future_to_actor[future_key] = (self._next_task_index, actor, cid)
+            self._next_task_index += 1
+
+            # update with future
+            self._cid_to_future[cid]["future"] = future_key
+
+    def submit_client_job(self, fn: Any, value: Callable, cid: str) -> None:
```

Comment:
Should this be the following?
```suggestion
    def submit_client_job(self, fn: Any, job_fn: Callable, cid: str) -> None:
```